feat: Failover2 plugin (#1089)
aws/aws-advanced-jdbc-wrapper: docs/using-the-jdbc-driver/using-plugins/UsingTheFailover2Plugin.md

<div style="text-align:center"><img src="../../images/failover2_panic.jpg" /></div>
<i>Picture 3. MonitoringRdsHostListProvider detects a new writer by establishing connections to nodes in separate threads.</i> 
<br><br>

When the cluster topology needs to be confirmed, the monitoring component opens new threads, one for each node (see _Picture 3_). Each of these threads tries to connect to a node and checks if the node is a writer. When Aurora failover occurs, the new writer node is the first node to reflect the true topology of the cluster. Other nodes connect to the new writer shortly after and update their local copies of the topology. Topology information acquired from a reader node may be outdated/inaccurate for a short period after failover. You can see a typical example of stale topology in the diagram above: thread `instance-3`, box `Topology`, to the right. The stale topology incorrectly shows that `instance-3` is still a writer.

> TThe threads monitoring the topology stop when a new writer is detected. For 30 seconds after a new writer is detected (and after all waiting connections have been notified), topology continues to be updated at an increased rate. This allows time for all readers to appear in the topology, since 30 seconds is usually enough time for cluster failover to complete and cluster topology to stabilize.

Reviewer: Extra T
```suggestion
The threads monitoring the topology stop when a new writer is detected. For 30 seconds after a new writer is detected (and after all waiting connections have been notified), topology continues to be updated at an increased rate. This allows time for all readers to appear in the topology, since 30 seconds is usually enough time for cluster failover to complete and cluster topology to stabilize.
```

Failover Plugin v2 uses an optimized approach when a process of detecting and confirming of a cluster topology is delegated to central topology monitoring component that runs in a separate thread. When topology is confirmed and a new writer is detected, each of waiting connection can resume and reconnect to a required node. Such design helps to minimize resources required for failover processing. It scales better comparing to Failover Plugin.

<div style="text-align:center"><img src="../../images/failover2.jpg" /></div>
<i>Picture 2. Connections call MonitoringRdsHostListProvider and delegate it a new writer detection. While waiting for MonitoringRdsHostListProvider, connection threads suspend.</i>
<br><br>

If two connections get communication issue with their internal (physical) DB connections, each connection may request a topology monitoring component (`MonitoringRdsHostListProvider` on a _Picture 2_) to provide a new writer. Both connections are notified as soon as a latest topology gets available. Connection threads can resume and continue with their suspended workflows and to reconnect to a reader or a writer node. 

> The topology monitoring component mentioned above (`MonitoringRdsHostListProvider`) updates topology periodically. Usually it uses a connection to a writer node to fetch a cluster topology. Using a connection to a writer node allows to get topology from the first hand without a risk of getting stale data as in case of fetching it from a reader. In some exceptional cases the monitoring component may (temporarily) use a reader connection to fetch a topology however it switch back to a writer node as soon as possible.

Reviewer 2: ```suggestion
The topology monitoring component mentioned above (`MonitoringRdsHostListProvider`) updates the topology periodically. Usually it uses a connection to a writer node to fetch the cluster topology. Using a connection to a writer node helps the component retrieve the topology without the risk of getting stale data, which may occur if fetching the topology from a reader. In some exceptional cases the monitoring component may temporarily use a reader connection to fetch the topology, but it will switch back to a writer node as soon as possible.
```

<div style="text-align:center"><img src="../../images/failover2_panic.jpg" /></div>
<i>Picture 3. MonitoringRdsHostListProvider detects a new writer by establishing connections to nodes in separate threads.</i> 
<br><br>

> When cluster topology needs to be confirmed, the monitoring component opens new threads, one for each node (see _Picture 3_). Each of these threads try to connect to a node and to check if node is a writer. Aurora clusters architected the way that a writer node gets updated first and provides a true topology. Other nodes connect to a new writer shortly after and update their local copies of topology. When reading topology from a reader node, it's quite possible to get a stale topology for some short time after failover. You can see a typical stale topology on a diagram above, thread `instance-3`, box `Topology` to the right. The topology shows that `instance-3` is still a writer there. 

Reviewer 2: ```suggestion
When the cluster topology needs to be confirmed, the monitoring component opens new threads, one for each node (see _Picture 3_). Each of these threads tries to connect to a node and checks if the node is a writer. Aurora clusters are architected in a way such that a writer node gets updated first and provides the true topology. Other nodes connect to a new writer shortly after and they update their local copies of the topology. When reading the topology from a reader node, it's quite possible to get a stale topology for a short time period after failover. You can see a typical stale topology on a diagram above, thread `instance-3`, box `Topology` to the right. The topology shows that `instance-3` is still a writer there. 
```

> Threads stop when a new writer is detected. 30s after a new writer is detected (and after all waiting connections have been notified), topology continues to be updated with a high rate. It guarantees that all readers appear in topology. Usually 30s is enough time to complete a cluster failover and to get a stable cluster topology. 

Reviewer 3: ```suggestion
Threads stop when a new writer is detected. 30s after a new writer is detected (and after all waiting connections have been notified), topology continues to be updated with a high rate. It guarantees that all readers will appear in the topology. Usually 30s is enough time to complete a cluster failover and to get a stable cluster topology. 
```

> | `clusterTopologyHighRefreshRateMs`    | Integer |                                        No                                        | Cluster topology refresh rate in milliseconds when a cluster is not in failover. It refers to `slow monitoring rate` as explained above.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | `30000`                                                                                                                                                                                             |

Author: It should be `clusterTopologyRefreshRateMs`. Thank you!
